core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java

```diff
@@ -201,6 +295,102 @@ static ExprValue getFormattedTime(ExprValue timeExpr, ExprValue formatExpr) {
     return getFormattedString(formatExpr, TIME_HANDLERS, time);
   }
 
+  private static ExprCoreType getReturnType(ExprValue formatStringExpr) {
+    String formatString = formatStringExpr.stringValue();
+
+    boolean hasDate = false;
+    boolean hasTime = false;
+
+    for (String dateSpecifier: DATE_SPECIFIERS) {
+      if (formatString.contains(dateSpecifier)){
+        hasDate = true;
+      }
+    }
+    for (String timeSpecifier: TIME_SPECIFIERS) {
+      if (formatString.contains(timeSpecifier)){
+        hasTime = true;
+      }
+    }
+
+    if (hasDate && hasTime){
+      return ExprCoreType.DATETIME;
+    } else if (hasTime) {
+      return ExprCoreType.TIME;
+    }
+
+    return ExprCoreType.DATE;
+  }
+
+  static private boolean checkValidTemporalAccessor(ExprCoreType type, TemporalAccessor ta) {
+    if (type == ExprCoreType.DATE) {
+      if (!ta.isSupported(ChronoField.YEAR) ||
+          !ta.isSupported(ChronoField.MONTH_OF_YEAR) ||
+          !ta.isSupported(ChronoField.DAY_OF_MONTH)) {
+        return false;
+      }
+    } else if (type == ExprCoreType.TIME) {
+      if (!ta.isSupported(ChronoField.HOUR_OF_DAY) ||
+          !ta.isSupported(ChronoField.MINUTE_OF_HOUR) ||
+          !ta.isSupported(ChronoField.SECOND_OF_MINUTE)) {
+        return false;
+      }
+    } else {
+      if (!ta.isSupported(ChronoField.YEAR) ||
+          !ta.isSupported(ChronoField.MONTH_OF_YEAR) ||
+          !ta.isSupported(ChronoField.DAY_OF_MONTH) ||
+          !ta.isSupported(ChronoField.HOUR_OF_DAY) ||
+          !ta.isSupported(ChronoField.MINUTE_OF_HOUR) ||
+          !ta.isSupported(ChronoField.SECOND_OF_MINUTE)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  static private ExprValue formatStrToDateReturn(TemporalAccessor ta, ExprValue formatExpr) {
+    switch (getReturnType(formatExpr)) {
+      case DATETIME:
+        return (checkValidTemporalAccessor(ExprCoreType.DATETIME, ta)) ?
+            new ExprDatetimeValue(LocalDateTime.from(ta)) : ExprNullValue.of();
+      case TIME:
+        return (checkValidTemporalAccessor(ExprCoreType.TIME, ta)) ?
+            new ExprTimeValue(LocalTime.from(ta)) : ExprNullValue.of();
+      default:
+        return (checkValidTemporalAccessor(ExprCoreType.DATE, ta)) ?
+            new ExprDateValue(LocalDate.from(ta)) : ExprNullValue.of();
```

Comment:
We have to return always DATETIME as declared in signature.

Reply:
I did it this way because of how it seems like `DATETIME` is defined in `ExprCoreType.java`. 

Is it possible to declare different return types for the same function? I originally had signatures for `(String,String) -> DATETIME`, `(String,String) -> DATE`, `(String,String) -> TIME` ...etc. to align with the [MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_str-to-date) docs, but was getting test failures because of how I was testing for the return type. Maybe it's better to always return a `DATETIME` instead of aligning with MySQL?

Reply:
MySQL returns different types, but we can't do that
```
mysql> select str_to_date('1984 02 15', '%Y %m %d'), str_to_date('10 20 30', '%H %i %s');
Field   1:  `str_to_date('1984 02 15', '%Y %m %d')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATE
Collation:  binary (63)
Length:     10
Max_length: 10
Decimals:   0
Flags:      BINARY

Field   2:  `str_to_date('10 20 30', '%H %i %s')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       TIME
Collation:  binary (63)
Length:     10
Max_length: 8
Decimals:   0
Flags:      BINARY


+---------------------------------------+-------------------------------------+
| str_to_date('1984 02 15', '%Y %m %d') | str_to_date('10 20 30', '%H %i %s') |
+---------------------------------------+-------------------------------------+
| 1984-02-15                            | 10:20:30                            |
+---------------------------------------+-------------------------------------+
1 row in set (0.00 sec)
```

The only function signature declares that it returns `DATETIME` values, so any other function receives that value should be able to get it by `value.datetimeValue()`.

Reply:
Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/231/commits/e431d471f257e51120c65d225c8fa05313e47316.